### ISSUE TITLE
[BUG] Exponential Smoothing produces different results across runs with pyspark UDFs

### ISSUE THREAD

**Issue by manuel-munoz-aguirre** [sktime/sktime#6175]
**Describe the bug**
While executing exponential smoothing in a parallel setting with pyspark using `applyInPandas`, the forecast output can vary across runs for the same input time series. It appears that it is possible to get (consistently) at least two distinct forecasts for the same input. I've been able to replicate this with different time series, but it does not happen all the time, which makes me believe it could be related to minimization within scipy or any other internal process in the smoother. This could potentially be related to [this issue](https://github.com/statsmodels/statsmodels/issues/8894), with the difference that all the machines in my compute cluster have the same configuration. This happens also with `method="L-BFGS-B"`, although with slighter variations.

The example below has a single dummy time series, forecasted `reps` times in parallel, and reproduces the variation in the output.

**To Reproduce**
```python
import numpy as np
import pandas as pd

from sktime.forecasting.exp_smoothing import ExponentialSmoothing
from sktime.forecasting.base import ForecastingHorizon

reps = 30
size = 52*3
val_size = 16
np.random.seed(100)

# Create groups for repetitions
groups = pd.DataFrame.from_records(zip(range(0, reps), [0]*reps), columns=["group", "key"])

# Create artificial data
y = pd.DataFrame(np.random.randint(size=size, low=1000, high=3000), columns=['y'])
dates = pd.date_range(start="2022-01-01", periods=size, freq="W-MON")
val_cutoff = dates[-val_size]
y['Period'] = dates.astype(str)
y['key'] = 0
y = (y
  .merge(groups, on="key", how="outer")
  .drop('key', axis=1)
  .sort_values('group')
  .reset_index(drop=True)
)

def es_parallel(data):
  i = data['group'].unique()[0]
  data['Period'] = pd.to_datetime(data['Period'])
  data = data.sort_values('Period', ascending=True)
  data = data.drop('group', axis=1)

  train = data[data['Period'] <= "2024-09-09"].set_index('Period')
  train.index = train.index.to_period()

  test = data[(data['Period'] > "2024-09-09")].set_index('Period')
  test.index = test.index.to_period()
  fh = ForecastingHorizon(test.index, is_relative=False)

  fit = ExponentialSmoothing(
    trend="add",
    sp=52,
    initialization_method="estimated",
    method=None,
    random_state=0
  ).fit(train)

  out = fit.predict(fh).reset_index(drop=False)
  out['index'] = out['index'].astype(str)
  out['group'] = i

  return out

output = (spark
  .createDataFrame(y)
  .groupby('group')
  .applyInPandas(es_parallel, schema="index string, y float, group int")
  .toPandas()
)

output.pivot(index="index", columns="group", values="y").head()
```

![output](https://github.com/sktime/sktime/assets/5576458/eb81930a-4ef7-449f-8b02-142b77dcef7f)

**Expected behavior**
Forecast output should be always equal across machines when the random seed is set.

**Additional context**
All machines have the same library versions installed. When running within a single machine, the output forecast is always consistent. 

**Versions**
System:
    python: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
   machine: Linux-5.15.0-1056-azure-x86_64-with-glibc2.35

Python dependencies:
          pip: 22.2.2
       sktime: 0.27.0
      sklearn: 1.1.1
       skbase: 0.7.5
        numpy: 1.21.5
        scipy: 1.9.1
       pandas: 1.4.4
   matplotlib: 3.5.2
       joblib: 1.2.0
        numba: 0.55.1
  statsmodels: 0.13.2
     pmdarima: 2.0.3
statsforecast: None
      tsfresh: None
      tslearn: None
        torch: 1.13.1+cpu
   tensorflow: 2.11.1
tensorflow_probability: None

**Comment by fkiraly** [comment]
`ExponentialSmoothing` is a direct interface to `statsmodels` `holt_winters`, and afaik `random_state` is directly passed on.

The situation is a bit complex due to `statsmodels`' inconsistent handling of random seeds. E.g., it is passed only in `predict` but not in `fit`, it is not a constructor parameter. So it is uncertain whether all randomness is controlled for, meant to be controlled for, and whether there is a bug.

That different machine setups lead to different results, whereas similar setups lead to the same results may indicate that the issue is in `statsmodels` or even deeper, e.g., `numpy.random` inconsistencies - I vaguely remember hardware specific inconsistencies.

So it might be worth investigating if it is `holt_winters`, or not.

**Comment by manuel-munoz-aguirre** [comment]
Sure, this runs without issues. The output is equal as in the toy example (and with the reproducibility issue too, as expected I guess). In the use case that I have for the parallelization though I need to stick to the UDF though for other reasons.

```python
from joblibspark import register_spark
register_spark()

y['Period'] = pd.to_datetime(y['Period'])
train_data = y[y['Period'] <= "2024-09-09"]
test_data = (
  y[(y['Period'] > "2024-09-09") & (y['group'] == 0)]
  .sort_values('Period', ascending=True)
  .set_index('Period')
)
test_data.index = test_data.index.to_period()
fh = ForecastingHorizon(test_data.index, is_relative=False)

f = ExponentialSmoothing(    
  trend="add",
  sp=52,
  initialization_method="estimated",
  method=None,
  random_state=0
)
f.set_config(**{
  "backend:parallel": "joblib",
  "backend:parallel:params": {'backend': 'spark'},
})

f.fit(train_data.set_index(['group', 'Period']).sort_index(), fh=fh)
result = f.predict()
```

Output:
![output_2](https://github.com/sktime/sktime/assets/5576458/f133d7a8-4ceb-4b2f-8b45-3976f8543d3a)

**Comment by fkiraly** [comment]
Yes, thanks. Indeed, your code should be using the `spark` backend.

The docs should be also updated - another user (@julnow) has already highlighted the lack of examples for multithreading, see here: https://github.com/sktime/sktime/issues/6174

**Comment by manuel-munoz-aguirre** [comment]
I've come to the conclusion that this behavior is due to the provisioned hardware in the spark compute cluster not being exactly identical in terms of CPU feature sets/SIMD extensions (even if machines belong to the same compute family). When the CPUs are exactly equal, this situation does not happen. Proposing to close as it is a hardware-related issue and not an sktime code issue.
